lib/tree.rb

```diff
@@ -2,59 +2,131 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) is the best case which is a balanced tree, where n is the number of nodes. Worst case would be O(n) if each nod was more than the previous, or all less than the previous, which would just be a linked list.
+  # Space Complexity:O(1) only constant sized variables are used
   def add(key, value)
-    raise NotImplementedError
+    node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = node
+      return
+    end
+    current = @root
+    until current == nil
+      if current.key > node.key
+        if current.left == nil
+          current.left = node
+          return
+        else
+          current = current.left
+        end
+      else
+        if current.right == nil
+          current.right = node
+          return
+        else
+          current = current.right
+        end
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) is the best case which is a balanced tree, where n is the number of nodes. Worst case would be O(n) if each nod was more than the previous, or all less than the previous, which would just be a linked list.
+  # Space Complexity:O(1) only constant sized variables are used
   def find(key)
-    raise NotImplementedError
+    current = @root
+    until current == nil
+      if current.key > key
+        current = current.left
+      elsif current.key < key
+        current = current.right
+      else
+        return current.value
+      end
+    end
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is the number of nodes. The inorder recurssive method is called once for each node, and the time complexity for each frame is constant.
+  # Space Complexity: Worst Case would be O(n^2) for a completely unbalanced tree, becuase you could have n stack frames and n arrays. Best case for a balances tree would be O(log n)^2 because you would have log n stack frames at once and log n arrays at once.
```

Comment:
O(n<sup>2</sup>)???  You are building an array of each key-value pair, and have the system stack being called on each node, but they're not nested.  Remember arrays are passed by **reference**.